File: R/rates.R
# --------------------------------------------------------------------------------
#   external forcing
# --------------------------------------------------------------------------------

#' @title Temperature
#' @description A modified cosine function.
#' @param t time
#' @param pars a [list]
#' @export
temperature <- function(t, pars){
  
  phi = pars[["phi"]] # PHASE
  lambda = pars[["lambda"]] # A
  mu = pars[["mu"]] # M
  gamma = pars[["gamma"]] # POWER
  
  temp = 0.0
  
  if(t < 0.0){
    temp = (mu - lambda) + lambda * 2.0 * (0.5 * (1.0 + cos(2.0 * pi * (0.0 - phi) / 365.0)))^gamma
  } else {
    temp = (mu - lambda) + lambda * 2.0 * (0.5 * (1.0 + cos(2.0 * pi * (t - phi) / 365.0)))^gamma
  }
  
  return(temp)
}

#' @title Photoperiod
#' @param t time
#' @param pars a [list]
#' @export
photoperiod <- function(t, pars){
  
  L = pars[["L"]] # latitude (51 in thesis)
  
  # define photoperiod values
  EPS = asin(0.39795 * cos(0.2163108 + 2 * atan(0.9671396 * tan(0.00860 * (t - 3.5)))))
  NUM = sin(0.8333 * pi/ 180.0) + (sin(L * pi / 180.0) * sin(EPS))
  DEN = cos(L * pi / 180.0) * cos(EPS)
  DAYLIGHT = 24.0 - (24.0 / pi) * acos(NUM / DEN)
  
  return(DAYLIGHT)
}


# --------------------------------------------------------------------------------
#   diapause
# --------------------------------------------------------------------------------

#' @title Diapause (spring)
#' @param pp photoperiod
#' @export
diapause_spring <- function(pp){
  1.0 / (1.0 + exp(5.0 * (14.0 - pp)))
}

#' @title Diapause (autumn)
#' @param pp photoperiod
#' @export
diapause_autumn <- function(pp){
  1.0 / (1.0 + exp(5.0 * (13.0 - pp)))
}


# --------------------------------------------------------------------------------
#   mortality
# --------------------------------------------------------------------------------

#' @title Egg mortality rate
#' @param temp temperature
#' @param pars a [list]
#' @export
death_egg_rate <- function(temp, pars){
  
  nu_0E = pars[["nu_0E"]] # U3
  nu_1E = pars[["nu_1E"]] # U4
  nu_2E = pars[["nu_2E"]] # U5
  death_max <- pars[["death_max"]]
  
  # calculate egg death rate
  egg_d = nu_0E * exp(((temp - nu_1E) / nu_2E)^2)
  
  if(egg_d > death_max){
    egg_d = death_max
  }
  
  return(egg_d)
}

#' @title Larvae mortality rate
#' @param temp temperature
#' @param pars a [list]
#' @export
death_larvae_rate <- function(temp, pars){
  
  nu_0L = pars[["nu_0L"]] # U3
  nu_1L = pars[["nu_1L"]] # U4
  nu_2L = pars[["nu_2L"]] # U5
  death_max <- pars[["death_max"]]
  
  # calculate egg death rate
  larvae_d = nu_0L * exp(((temp - nu_1L) / nu_2L)^2)
  
  if(larvae_d > death_max){
    larvae_d = death_max
  }
  
  return(larvae_d)
}

#' @title Pupae mortality rate
#' @param temp temperature
#' @param pars a [list]
#' @export
death_pupae_rate <- function(temp, pars){
  
  nu_0P = pars[["nu_0P"]] # U3
  nu_1P = pars[["nu_1P"]] # U4
  nu_2P = pars[["nu_2P"]] # U5
  death_max <- pars[["death_max"]]
  
  # calculate egg death rate
  pupal_d = nu_0P * exp(((temp - nu_1P)/nu_2P)^2)
  
  if(pupal_d > death_max){
    pupal_d = death_max
  }
  return(pupal_d)
}

#' @title Adult mortality rate
#' @param temp temperature
#' @param pars a [list]
#' @export
death_adult_rate <- function(temp, pars){
  
  alpha_A = pars[["alpha_A"]] # ALPHA
  beta_A = pars[["beta_A"]] # BETA
  death_min_a = pars[["death_min_a"]]
  
  # calculate adult death rate
  adult_d = alpha_A * (temp^beta_A)
  
  if(adult_d < death_min_a){
    adult_d = death_min_a
  }
  return(adult_d)
}


# --------------------------------------------------------------------------------
#   lifecycle stage progression rates
# --------------------------------------------------------------------------------

#' @title Gonotrophic cycle length
#' @param temp temperature
#' @param pars a [list]
#' @export
gonotrophic <- function(temp, pars){
  
  q1 = pars[["q1"]] # KG
  q2 = pars[["q2"]] # QG
  q3 = pars[["q3"]] # BG
  gon_min = pars[["gon_min"]]
  
  # calculate gonotrophic cycle length
  if(temp < 0.0){
    grate = 0.0333
  } else {
    grate = q1 / (1 + q2*exp(-q3*temp))
  }
  
  if(grate < gon_min){
    grate = gon_min
  }
  
  return(1.0 / grate)
}

#' @title Egg laying rate
#' @param d diapause
#' @param G duration of gonotrophic cycle
#' @param pars a [list]
#' @export
oviposition <- function(d, G, pars){
  
  max_egg = pars[["max_egg"]]
  
  egg_raft = d * max_egg * 0.5
  ovi = egg_raft / G
  
  return(ovi)
}

#' @title Egg maturation rate
#' @param temp temperature
#' @param pars a list
#' @export
egg_maturation_rate <- function(temp, pars){
  
  alpha_E = pars[["alpha_E"]] # ALPHA
  beta_E = pars[["beta_E"]] # BETA
  maturation_min = pars[["maturation_min"]]
  
  # calculate egg development rate
  if(temp < 0.0){
    egg_maturation = 0.016667
  } else {
    egg_maturation = alpha_E * (temp^beta_E)
  }
  
  if(egg_maturation < maturation_min){
    egg_maturation = maturation_min
  }
  return(egg_maturation)
}

#' @title Larvae maturation rate
#' @param temp temperature
#' @param pars a list
#' @export
larvae_maturation_rate <- function(temp, pars){
  
  alpha_L = pars[["alpha_L"]] # ALPHA
  beta_L = pars[["beta_L"]] # BETA
  maturation_min = pars[["maturation_min"]]
  
  # calculate larvae development rate
  if(temp < 0.0){
    larvae_maturation = 0.016667
  } else {
    larvae_maturation = alpha_L * (temp^beta_L)
  }
  
  if(larvae_maturation < maturation_min){
    larvae_maturation = maturation_min
  }
  return(larvae_maturation)
}

#' @title Pupae maturation rate
#' @param temp temperature
#' @param pars a list
#' @export
pupae_maturation_rate <- function(temp, pars){
  
  alpha_P = pars[["alpha_P"]] # ALPHA
  beta_P = pars[["beta_P"]] # BETA
  maturation_min = pars[["maturation_min"]]
  
  # calculate larvae development rate
  if(temp < 0.0){
    pupae_maturation = 0.016667
  } else {
    pupae_maturation = alpha_P * (temp^beta_P)
  }
  
  if(pupae_maturation < maturation_min){
    pupae_maturation = maturation_min
  }
  return(pupae_maturation)
}


# --------------------------------------------------------------------------------
#   extrinsic incubation period
# --------------------------------------------------------------------------------
  
#' @title Extrinsic incubation development rate
#' @param temp temperature
#' @param pars a list
#' @export
eip_rate <- function(temp, pars) {
    
  q = pars[["eip_q"]]
  Tmax = pars[["eip_tmax"]]
  Tmin = pars[["eip_tmin"]]
  eip_min = pars[["eip_min"]]
    
  if (temp > Tmax) {
    return(eip_min)
  } else if (temp < Tmin) {
    return(eip_min)
  } else {
    eip = q*temp*(temp - Tmin) * sqrt(Tmax - temp)
    return(max(eip, eip_min))
  }
    
}

File: src/rates.cpp
#include "rates.h"


// --------------------------------------------------------------------------------
//   mortality
// --------------------------------------------------------------------------------

// for each life stage, mortality is computed; returns vector of rates, input is temperature
mortality_rate_fn make_egg_mortality_rate_ewing(const Rcpp::List& pars) {
  double nu_0E = Rcpp::as<double>(pars["nu_0E"]); // U3
  double nu_1E = Rcpp::as<double>(pars["nu_1E"]); // U4
  double nu_2E = Rcpp::as<double>(pars["nu_2E"]); // U5
  double death_max = Rcpp::as<double>(pars["death_max"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [nu_0E, nu_1E, nu_2E, death_max, n_patch] (const arma::Col<double>& temp) -> arma::Col<double> {
    arma::Col<double> mort(n_patch, arma::fill::zeros);
    for (auto i = 0u; i < n_patch; ++i) {
      mort(i) = std::min(nu_0E * exp(pow((temp(i) - nu_1E) / nu_2E, 2.0)), death_max);
    }
    return mort;
  };
};

mortality_rate_fn make_larvae_mortality_rate_ewing(const Rcpp::List& pars) {
  double nu_0L = Rcpp::as<double>(pars["nu_0L"]); // U3
  double nu_1L = Rcpp::as<double>(pars["nu_1L"]); // U4
  double nu_2L = Rcpp::as<double>(pars["nu_2L"]); // U5
  double death_max = Rcpp::as<double>(pars["death_max"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [nu_0L, nu_1L, nu_2L, death_max, n_patch] (const arma::Col<double>& temp) -> arma::Col<double> {
    arma::Col<double> mort(n_patch, arma::fill::zeros);
    for (auto i = 0u; i < n_patch; ++i) {
      mort(i) = std::min(nu_0L * exp(pow((temp(i) - nu_1L) / nu_2L, 2.0)), death_max);
    }
    return mort;
  };
};

mortality_rate_fn make_pupae_mortality_rate_ewing(const Rcpp::List& pars) {
  double nu_0P = Rcpp::as<double>(pars["nu_0P"]); // U3
  double nu_1P = Rcpp::as<double>(pars["nu_1P"]); // U4
  double nu_2P = Rcpp::as<double>(pars["nu_2P"]); // U5
  double death_max = Rcpp::as<double>(pars["death_max"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [nu_0P, nu_1P, nu_2P, death_max, n_patch] (const arma::Col<double>& temp) -> arma::Col<double> {
    arma::Col<double> mort(n_patch, arma::fill::zeros);
    for (auto i = 0u; i < n_patch; ++i) {
      mort(i) = std::min(nu_0P * exp(pow((temp(i) - nu_1P) / nu_2P, 2.0)), death_max);
    }
    return mort;
  };
};

mortality_rate_fn make_adult_mortality_rate_ewing(const Rcpp::List& pars) {
  double alpha_A = Rcpp::as<double>(pars["alpha_A"]); // ALPHA
  double beta_A = Rcpp::as<double>(pars["beta_A"]); // # BETA
  double death_min_a = Rcpp::as<double>(pars["death_min_a"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [alpha_A, beta_A, death_min_a, n_patch] (const arma::Col<double>& temp) -> arma::Col<double> {
    arma::Col<double> mort(n_patch, arma::fill::zeros);
    for (auto i = 0u; i < n_patch; ++i) {
      mort(i) = std::max(alpha_A * pow(temp(i), beta_A), death_min_a);
    }
    return mort;
  };
};


// --------------------------------------------------------------------------------
//   gonotrophic rate
// --------------------------------------------------------------------------------

gonotrophic_fn make_gonotrophic_ewing(const Rcpp::List& pars) {
  double q1 = Rcpp::as<double>(pars["q1"]); // KG
  double q2 = Rcpp::as<double>(pars["q2"]); // QG
  double q3 = Rcpp::as<double>(pars["q3"]); // BG
  double gon_zero_temp = Rcpp::as<double>(pars["gon_zero_temp"]);
  double gon_min = Rcpp::as<double>(pars["gon_min"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [q1, q2, q3, gon_min, gon_zero_temp, n_patch] (const arma::Col<double>& temp, const arma::Col<double>& diapause) -> arma::Col<double> {
    arma::Col<double> gon(n_patch, arma::fill::zeros);
    for (auto i = 0u; i < n_patch; ++i) {
      if (temp(i) < 0.0) {
        gon(i) = gon_zero_temp;
      } else {
        gon(i) = q1 / (1.0 + q2*exp(-q3*temp(i)));
      }
      
      if (gon(i) < gon_min) {
        gon(i) = gon_min;
      }
      
      gon(i) *= diapause(i);
    }
    return gon;
  };
};


// --------------------------------------------------------------------------------
//   oviposition
// --------------------------------------------------------------------------------

oviposition_fn make_oviposition_ewing(const Rcpp::List& pars) {
  double max_egg = Rcpp::as<double>(pars["max_egg"]);
  size_t n_patch = Rcpp::as<size_t>(pars["n_patch"]);
  
  return [max_egg, n_patch] (const arma::Col<double>& gonotrophic) -> arma::Col<double> {
    return gonotrophic * max_egg * 0.5;
  };
};